pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+		let mut num_leaving = 0;
+
+		transactional::with_storage_layer(|| -> DispatchResult {
+			NextEpochChanges::<T>::try_mutate(|changes| -> DispatchResult {
+				ActiveEpochData::<T>::try_mutate(|epoch_data| {
+					num_joining = changes.collators.inc.len();
+					num_leaving = changes.collators.out.len();
+					for leaving in changes.collators.out.drain(..) {
+						Self::do_exit_collator(&leaving)?;
+					}
+					for joining in changes.collators.inc.drain(..) {
+						Self::do_init_collator(&joining)?;
+					}
+
+					groups = T::Rewards::distribute_reward_with_weights(
+						epoch_data.reward,
+						epoch_data.weights.iter().map(|(g, w)| (*g, *w)),
+					)
+					.map(|results| {
+						Self::do_mint_remainder(epoch_data.reward, &results)?;
+						Ok(results.len() as u32)
+					})
+					.and_then(|num_groups| num_groups)?;
+
+					for (&group_id, &weight) in &changes.weights {
+						epoch_data.weights.try_insert(group_id, weight).ok();
+						weight_changes += 1;
+					}
+
+					epoch_data.reward = changes.reward.unwrap_or(epoch_data.reward);
+
+					Self::deposit_event(Event::NewEpoch {
+						reward: epoch_data.reward,
+						last_changes: mem::take(changes),
+					});
+
+					Ok(())
+				})
+			})
+		})
+		.ok();
+
+		// TODO: Apply number of incoming and outgoing collators
+		T::WeightInfo::on_initialize(groups, weight_changes)
+	}
+}
+
+impl<T: Config> sp_runtime::BoundToRuntimeAppPublic for Pallet<T> {
+	type Public = T::AuthorityId;
+}
+
+// Should be instantiated after the original SessionHandler such that current and queued collators are up-to-date for the current session.
+impl<T: Config> OneSessionHandler<T::AccountId> for Pallet<T> {
+	type Key = T::AuthorityId;
+
+	fn on_genesis_session<'a, I: 'a>(_validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+	{
+		// we don't care.
+	}
+
+	fn on_new_session<'a, I: 'a>(changed: bool, validators: I, queued_validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+	{
+		// MUST be called before updating collator set changes.
+		// Else the timing is off.
+		let mut weight = Self::do_advance_epoch();
+
+		// Prepare collator set changes for next session.
+		if changed {
+			let current = validators
+				.map(|(acc_id, _)| acc_id.clone())
+				.collect::<Vec<_>>();
+			let next = queued_validators
+				.map(|(acc_id, _)| acc_id.clone())
+				.collect::<Vec<_>>();
+
+			// Prepare for next session
+			if current != next {
+				NextEpochChanges::<T>::mutate(|EpochChanges { collators, .. }| {
+					let inc = next
+						.clone()
+						.into_iter()
+						.filter(|n| !current.iter().any(|curr| curr == n))
+						.collect::<Vec<_>>();
+					let out = current
+						.clone()
+						.into_iter()
+						.filter(|curr| !next.iter().any(|n| n == curr))
+						.collect::<Vec<_>>();
+					collators.inc = BoundedVec::<_, T::MaxCollators>::truncate_from(inc);
+					collators.out = BoundedVec::<_, T::MaxCollators>::truncate_from(out);
```

Comment:
What would happen with the currency minted by the collators that don't fit into the vec? (Could this even happen?)
Should we call `do_exit_collator()` for those?

Reply:
> What would happen with the currency minted by the collators that don't fit into the vec? (Could this even happen?)

The minting does not happen here. Here, we just note an upcoming collator set change in the next session/epoch `i+1`. Then, in the next epoch `i+1`, we reward these collators. 

Say we have more collators than allowed by our pallets bound, then the overflowing ones don't receive rewards. But since `MaxCollators` should be configured as `pallet_collator_selection::Config::MaxCandidates`, this is impossible to happen.
I will look into utilizing `integrity-tests` to ensure such assumptions hold true at compile time. Haven't used that hook before but recently stumbled upon it.

Reply:
You're right regarding the collator, I saw the `do_advance_epoch()` on the top, but it uses the previous ones.

> I will look into utilizing integrity-tests to ensure such assumptions hold true at compile time. Haven't used that hook before but recently stumbled upon it.

I'll take a look. Sounds interesting, thanks!